async_generator/_impl.py

```diff
@@ -263,47 +310,59 @@ def ag_frame(self):
     # Core functionality
     ################################################################
 
-    # We make these async functions and use await, rather than just regular
-    # functions that pass back awaitables, in order to get more useful
-    # tracebacks when debugging.
+    # These need to return awaitables, rather than being async functions,
+    # to match the native behavior where the firstiter hook is called
+    # immediately on asend()/etc, even if the coroutine that asend()
+    # produces isn't awaited for a bit.
+
+    def __anext__(self):
+        return self._do_it(self._it.__next__)
 
-    async def __anext__(self):
-        return await self._do_it(self._it.__next__)
+    def asend(self, value):
+        return self._do_it(self._it.send, value)
 
-    async def asend(self, value):
-        return await self._do_it(self._it.send, value)
+    def athrow(self, type, value=None, traceback=None):
+        return self._do_it(self._it.throw, type, value, traceback)
 
-    async def athrow(self, type, value=None, traceback=None):
-        return await self._do_it(self._it.throw, type, value, traceback)
+    def _do_it(self, start_fn, *args):
+        coro_state = getcoroutinestate(self._coroutine)
+        if coro_state is CORO_CREATED:
+            (firstiter, self._finalizer) = get_asyncgen_hooks()
+            if firstiter is not None:
+                firstiter(self)
```

Comment:
This could call the `firstiter` hook twice:

```python
agen = agen_function()
agen.asend(None).close()
agen.asend(None).close()
```
Pretty obscure, but native agens don't seem to call it twice in this case. We should add a test case, and I guess an explicit have-we-done-hook-setup variable.

Reply:
Thanks for spotting! Fixed and added a test.